File: Graphs/1697. Checking Existence of Edge Length Limited Paths.py
class Solution:
    def distanceLimitedPathsExist(self, n: int, edgeList: List[List[int]], queries: List[List[int]]) -> List[bool]:
        p  = len(edgeList)
        q = len(queries)
        parents = [i for i in range(n)]
        def find(node):
            if parents[node] != node:
                parents[node] = find(parents[node])
            return parents[node]

        def union(n1, n2):
            p1, p2 = find(n1), find(n2)
            if p1 == p2:
                return
            parents[p2] = p1
            return

        res = [False for i in range(q)]

        queries_new = [(a, b, c, d) for d, (a, b, c) in enumerate(queries)]
        queries_new.sort(key = lambda x: x[2]) # sort according to the weight
        edgeList.sort(key = lambda x: x[2])
        ptr1 = 0
        ptr2 = 0

        for query in queries_new:
            while ptr1 < len(edgeList) and edgeList[ptr1][2] < query[2]:
                union(edgeList[ptr1][0], edgeList[ptr1][1])
                ptr1 += 1
            res[query[3]] = (find(query[0]) == find(query[1]))
        return res

File: Graphs/1697. Checking Existence of Edge Length Limited Paths.cpp
class Solution {
public:
    bool static compare1(vector<int>& a, vector<int>& b) {
        return a[3] < b[3];
    }
    bool static compare2(vector<int>& a, vector<int>& b) {
        return a[2] < b[2];
    }

    int find(int node, vector<int>& parents) {
        if (parents[node] != node) {
            parents[node] = find(parents[node], parents);
        }

        return parents[node];

    }
    void u(int n1, int n2, vector<int>& parents) {
        int p1 = find(n1, parents);
        int p2 = find(n2, parents);

        if (p1 == p2) {
            return;
        }

        parents[p2] = p1;
        return;
    }


    vector<bool> distanceLimitedPathsExist(int n, vector<vector<int>>& edgeList, vector<vector<int>>& queries) {
        vector<int> parents(n, 0);
        vector<vector<int>> newQueries;
        vector<bool> res(queries.size(), false);

        for (int i = 0; i < n; i++) {
            parents[i] = i;
        }
        for (int i = 0; i < queries.size(); i++) {
            newQueries.push_back({queries[i][0], queries[i][1], queries[i][2], i});
        }

        sort(newQueries.begin(), newQueries.end(), compare2);
        sort(edgeList.begin(), edgeList.end(), compare2);

        int ptr1 = 0 ;
        for (int i = 0; i < newQueries.size(); i++) {
            while (ptr1 < edgeList.size() && edgeList[ptr1][2] < newQueries[i][2]) {
                u(edgeList[ptr1][0], edgeList[ptr1][1], parents);
                ptr1++;
            }

            res[newQueries[i][3]] = find(newQueries[i][0], parents) == find(newQueries[i][1], parents);

        }

        return res;


    }




};